### ISSUE TITLE
Parsing suggestions for various devices

### ISSUE THREAD

**Issue by benbeckingsale** [nielsbasjes/yauaa#210]
Hi Niels, I have found a few prevalent useragents whose parsings could potentially be improved. I have detailed them below along with test cases. Would you be able to guide me on how best to add these to your library?

### First-Gen Amazon Echos
    Echo/1.0(APNG)

UAs from these devices are being classified as ‘Unknown’ in the device fields. Is it possible to update the matching as below?

    DeviceClass: Unknown  -->  Voice
    DeviceBrand: Unknown  -->  Amazon
    DeviceName:  Unknown  -->  Amazon Echo

Are you able to parse 1.0 as the Device Version?

### Sonos Speakers
    Linux UPnP/1.0 Sonos/53.3-71050 (ZPS1)
    Linux UPnP/1.0 Sonos/53.3-71050 (ZPS12)
    Linux UPnP/1.0 Sonos/53.3-71050 (ZP120)

UAs like these are also being classified as ‘Unknown’ in the device fields. I'm confident that `(ZPS1)` refers to the device model. Could you update the matching so that:

    DeviceBrand: Unknown  -->  Sonos
    DeviceName:  Unknown  -->  Sonos ZPS1

Speakers like this don’t seem to fit anywhere in YAUAA’s range of device classes. Would you consider introducing a dedicated class, such as ‘Smart Speaker’ for this kind of device?

### Playstation 3 Consoles
    PS3Application libhttp/4.8.5-000 (CellOS)

These UAs are also being classified as ‘Unknown’ in the device fields. Could you update the matching so that:

    DeviceClass: Unknown  -->  Game Console
    DeviceBrand: Unknown  -->  Sony
    DeviceName:  Unknown  -->  Playstation 3
    OSName:      Unknown  -->  CellOS

### Spotify Mobile App
    (1) Spotify/8.5.33 Android/28 (SM-G950F)
        Spotify/8.5.33 Android/29 (Pixel 3a)
    (2) Spotify/112000510 Win32/0 (PC laptop)
    (3) Spotify/8.5.33 iOS/13.1.3 (iPhone11,8)

UAs from the Android Spotify app `(1)` are being parsed as:

    DeviceClass: Mobile
    DeviceBrand: Unknown
    DeviceName:  Android Mobile
    AgentClass:  Browser
    AgentName:   Stock Android Browser

Similarly, with Windows `(2)`:

    DeviceClass: Unknown
    DeviceBrand: Unknown
    DeviceName:  Unknown
    AgentName:   Spotify

However, equivalent UAs from iOS devices `(3)` are parsed successfully as:

    DeviceClass: (Phone|Tablet)
    DeviceBrand: Apple
    DeviceName:  Apple (iPhone|iPad)
    AgentClass:  Special
    AgentName:   Spotify

I also noticed that the device name is usually found at the end of the Spotify UA (e.g. SM-G950F). Is it possible to parse this into `Device Name`?

Altogether, the ideal output for UA `(1)` would be:

    DeviceClass: Mobile
    DeviceBrand: Unknown                --> Samsung / Google
    DeviceName:  Android Mobile         --> Samsung SM-G950F / Google Pixel 3a
    AgentClass:  Browser                --> Special
    AgentName:   Stock Android Browser  --> Spotify

and for UA `(2)`

    DeviceClass: Unknown  -->  Desktop
    DeviceBrand: Unknown
    DeviceName:  Unknown
    OSName:      Unknown  -->  Windows
    AgentName:   Spotify  -->  Spotify  

## Test cases

1. Echo useragents
```
- test:
    input:
      user_agent_string: 'Echo/1.0(APNG)'
    expected:
      DeviceClass                          : 'Voice'
      DeviceName                           : 'Amazon Echo'
      DeviceBrand                          : 'Amazon'
      OperatingSystemClass                 : 'Mobile'
      OperatingSystemName                  : 'FireOS'
      OperatingSystemVersion               : '??'
      OperatingSystemVersionMajor          : '??'
      OperatingSystemNameVersion           : 'FireOS ??'
      OperatingSystemNameVersionMajor      : 'FireOS ??'
      OperatingSystemVersionBuild          : '??'
      LayoutEngineClass                    : 'Unknown'
      LayoutEdngineName                    : 'Unknown'
      LayoutEngineVersion                  : '??'
      LayoutEngineVersionMajor             : '??'
      LayoutEngineNameVersion              : 'Unknown ??'
      LayoutEngineNameVersionMajor         : 'Unknown ??'
      AgentClass                           : 'Special'
      AgentName                            : 'Echo'
      AgentVersion                         : '1.0'
      AgentVersionMajor                    : '1'
      AgentNameVersion                     : 'Echo 1.0'
      AgentNameVersionMajor                : 'Echo 1'
```

2. Sonos useragents
```
- test:
    input:
      user_agent_string: 'Linux UPnP/1.0 Sonos/53.3-71050 (ZPS1)'
    expected:
      DeviceClass: 'Smart Speaker'
      DeviceName: 'Sonos ZPS1'
      DeviceBrand: 'Sonos'
      OperatingSystemClass: 'Embedded'
      OperatingSystemName: 'Unknown'
      OperatingSystemVersion: '??'
      OperatingSystemVersionMajor: '??'
      OperatingSystemNameVersion: 'Unknown ??'
      OperatingSystemNameVersionMajor: 'Unknown ??'
      OperatingSystemVersionBuild: '??'
      LayoutEngineClass: 'Unknown'
      LayoutEngineName: 'Unknown'
      LayoutEngineVersion: '??'
      LayoutEngineVersionMajor: '??'
      LayoutEngineNameVersion: 'Unknown ??'
      LayoutEngineNameVersionMajor: 'Unknown ??'
      AgentClass: 'Special'
      AgentName: 'Linux UPnP'
      AgentVersion: '1.0'
      AgentVersionMajor: '1'
      AgentNameVersion: 'Linux UPnP 1.0'
      AgentNameVersionMajor: 'Linux UPnP 1'
```

3. Playstation 3 useragents
```
- test:
  input:
    user_agent_string:'PS3Application libhttp/4.8.5-000 (CellOS)'
  expected:
    DeviceClass: 'Game Console'
    DeviceName: 'Sony PlayStation 3'
    DeviceBrand: 'Sony'
    DeviceVersion: 'Unknown ??'
    OperatingSystemClass: 'Game Console'
    OperatingSystemName: 'CellOS'
    OperatingSystemVersion: '??'
    OperatingSystemVersionMajor: '??'
    OperatingSystemNameVersion: 'CellOS ??'
    OperatingSystemNameVersionMajor: 'CellOS ??'
    LayoutEngineClass: 'Unknown'
    LayoutEngineName: 'Unknown'
    LayoutEngineVersion: '??'
    LayoutEngineVersionMajor: '??'
    LayoutEngineNameVersion: 'Unknown ??'
    LayoutEngineNameVersionMajor: 'Unknown ??'
    AgentClass: 'Special'
    AgentName: 'PS3Application libhttp'
    AgentVersion: '4.8.5-000'
    AgentVersionMajor: '4'
    AgentNameVersion: 'PS3Applicationlibhttp 4.8.5-000'
    AgentNameVersionMajor: 'PS3Applicationlibhttp 4'
```

4. Spotify useragents
    (I inferred Android 9.0 from `/28` according to https://en.wikipedia.org/wiki/Android_version_history#Android_9_Pie_(API_28)).
```
- test:
  input:
    user_agent_string: 'Spotify/8.5.33 Android/28 (SM-G950F)'
  expected:
    DeviceClass: 'Phone'
    DeviceName: 'Samsung SM-G950F'
    DeviceBrand: 'Samsung'
    OperatingSystemClass: 'Mobile'
    OperatingSystemName: 'Android'
    OperatingSystemVersion: '9'
    OperatingSystemVersionMajor: '9'
    OperatingSystemNameVersion: 'Android 9'
    OperatingSystemNameVersionMajor: 'Android 9'
    OperatingSystemVersionBuild: 'Unknown'
    LayoutEngineClass: 'Unknown'
    LayoutEngineName: 'Unknown'
    LayoutEngineVersion: '??'
    LayoutEngineVersionMajor: '??'
    LayoutEngineNameVersion: 'UCBrowser U3 0.8.0'
    LayoutEngineNameVersionMajor: 'UCBrowser U3 0'
    AgentClass: 'Special'
    AgentName: 'Spotify'
    AgentVersion: '8.5.33'
    AgentVersionMajor: '8'
    AgentNameVersion: 'Spotify 8.5.33'
    AgentNameVersionMajor: 'Spotify 8'

- test:
  input:
    user_agent_string: 'Spotify/8.5.33 Android/29 (Pixel 3a)'
  expected:
    DeviceClass: 'Phone'
    DeviceName: 'Google Pixel 3a'
    DeviceBrand: 'Google'
    OperatingSystemClass: 'Mobile'
    OperatingSystemName: 'Android'
    OperatingSystemVersion: '10'
    OperatingSystemVersionMajor: '10'
    OperatingSystemNameVersion: 'Android 10'
    OperatingSystemNameVersionMajor: 'Android 10'
    OperatingSystemVersionBuild: 'Unknown'
    LayoutEngineClass: 'Unknown'
    LayoutEngineName: 'Unknown'
    LayoutEngineVersion: '??'
    LayoutEngineVersionMajor: '??'
    LayoutEngineNameVersion: 'UCBrowser U3 0.8.0'
    LayoutEngineNameVersionMajor: 'UCBrowser U3 0'
    AgentClass: 'Special'
    AgentName: 'Spotify'
    AgentVersion: '8.5.33'
    AgentVersionMajor: '8'
    AgentNameVersion: 'Spotify 8.5.33'
    AgentNameVersionMajor: 'Spotify 8'

- test:
  input:
    user_agent_string: 'Spotify/112000510 Win32/0 (PC laptop)'
  expected:
    DeviceClass: 'Desktop'
    DeviceName: 'Unknown'
    DeviceBrand: 'Unknown'
    OperatingSystemClass: 'Desktop'
    OperatingSystemName: 'Windows'
    OperatingSystemVersion: '??'
    OperatingSystemVersionMajor: '??'
    OperatingSystemNameVersion: 'Windows ??'
    OperatingSystemNameVersionMajor: 'Windows ??'
    OperatingSystemVersionBuild: 'Unknown'
    LayoutEngineClass: 'Unknown'
    LayoutEngineName: 'Unknown'
    LayoutEngineVersion: '??'
    LayoutEngineVersionMajor: '??'
    LayoutEngineNameVersion: 'UCBrowser U3 0.8.0'
    LayoutEngineNameVersionMajor: 'UCBrowser U3 0'
    AgentClass: 'Special'
    AgentName: 'Spotify'
    AgentVersion: '112000510'
    AgentVersionMajor: '112000510'
    AgentNameVersion: 'Spotify 112000510'
    AgentNameVersionMajor: 'Spotify 112000510'
```

**Comment by nielsbasjes** [comment]
Hi,
You put up quite a list. Thanks!
I'm looking into them and I'm adding most of them.

The PS3 example is very different from what I have seen from my own PS3. Where does it come from?

Seems the  `LayoutEngineNameVersion: 'UCBrowser U3 0.8.0'` and related fields have been copied a few times too many in your tests. So I'm ignoring those.

I'm curious; In some of those clients I have the question where these useragents appear?
Where in your logfiles does a Sonos speaker leave it's marks? Same for the Alexa, Echo and Spotify devices. Without any knowledge I would expect these systems to only appear in the logfiles of the companies that created these devices.

**Comment by nielsbasjes** [comment]
You write about the Sonos devices: `Speakers like this don’t seem to fit anywhere in YAUAA’s range of device classes. Would you consider introducing a dedicated class, such as ‘Smart Speaker’ for this kind of device?`

I do not know these devices. What makes them special?
Aren't all of those voice devices (Alexa, Google Home, etc.) in reality "Smart Speakers" ?

**Comment by nielsbasjes** [comment]
I changed the outcome for Sonos from what you proposed
```
- test:
    input:
      user_agent_string: 'Linux UPnP/1.0 Sonos/53.3-71050 (ZPS1)'
    expected:
      DeviceClass                          : 'Voice'
      DeviceName                           : 'Sonos ZPS1'
      DeviceBrand                          : 'Sonos'
      OperatingSystemClass                 : 'Embedded'
      OperatingSystemName                  : 'Unknown'
      OperatingSystemVersion               : '??'
      OperatingSystemVersionMajor          : '??'
      OperatingSystemNameVersion           : 'Unknown ??'
      OperatingSystemNameVersionMajor      : 'Unknown ??'
      OperatingSystemVersionBuild          : '??'
      LayoutEngineClass                    : 'Unknown'
      LayoutEngineName                     : 'Unknown'
      LayoutEngineVersion                  : '??'
      LayoutEngineVersionMajor             : '??'
      LayoutEngineNameVersion              : 'Unknown ??'
      LayoutEngineNameVersionMajor         : 'Unknown ??'
      AgentClass                           : 'Special'
      AgentName                            : 'Linux UPnP'
      AgentVersion                         : '1.0'
      AgentVersionMajor                    : '1'
      AgentNameVersion                     : 'Linux UPnP 1.0'
      AgentNameVersionMajor                : 'Linux UPnP 1'
```

into this

```
- test:
    input:
      user_agent_string: 'Linux UPnP/1.0 Sonos/53.3-71050 (ZPS1)'
    expected:
      DeviceClass                          : 'Voice'
      DeviceName                           : 'Sonos ZPS1'
      DeviceBrand                          : 'Sonos'
      OperatingSystemClass                 : 'Embedded'
      OperatingSystemName                  : 'Linux'
      OperatingSystemVersion               : '??'
      OperatingSystemVersionMajor          : '??'
      OperatingSystemNameVersion           : 'Linux ??'
      OperatingSystemNameVersionMajor      : 'Linux ??'
      OperatingSystemVersionBuild          : '??'
      LayoutEngineClass                    : 'Unknown'
      LayoutEngineName                     : 'Unknown'
      LayoutEngineVersion                  : '??'
      LayoutEngineVersionMajor             : '??'
      LayoutEngineNameVersion              : 'Unknown ??'
      LayoutEngineNameVersionMajor         : 'Unknown ??'
      AgentClass                           : 'Special'
      AgentName                            : 'Sonos'
      AgentVersion                         : '53.3-71050'
      AgentVersionMajor                    : '53'
      AgentNameVersion                     : 'Sonos 53.3-71050'
      AgentNameVersionMajor                : 'Sonos 53'

```

**Comment by nielsbasjes** [comment]
@benbeckingsale Can you please have a look at the test cases in https://github.com/nielsbasjes/yauaa/blob/AudioDevices/analyzer/src/main/resources/UserAgents/Audio.yaml ?
These are most of the changes you proposed after I tweaked them.
Are these now correct?

**Comment by benbeckingsale** [comment]
Thanks for taking the time to address each suggestion.

> I have the question where these useragents appear?

I am using YAUAA for media logs rather than website/HTTP logs, so some useragents will reflect a device's lower-level media API.

### Sonos

The change from `Unknown` to `Linux` looks good – thanks.

[This site](https://www.imore.com/how-does-sonos-work) describes Sonos as

> a system of wireless speakers that connect together so you can stream music and even set up a home theater system in any room or every room, and control it all with your phone or tablet.

The speakers that you mentioned are indeed "smart speakers", and some Sonos ones are, too.

However, I'm not confident that a Sonos speaker would be a "voice" device. I would say that `DeviceClass: Unknown` is a more accurate statement, as I can't say for sure whether the useragent comes from the speaker itself or a phone/tablet acting as a remote, etc.

So, would you be happy to change `DeviceClass: Voice  -->  Unknown` for the Sonos test cases?

Everything else looks good!

**Comment by nielsbasjes** [comment]
This is a class of devices that seems to evolve very quickly in what they can do.
Right now I'm considering to put all of those devices under the DeviceClass "Voice".

That would effectively make this category "Everything that is mainly focused around sound".
So that includes things that listen like Alexa, Google Home and also systems like the Sonos Speakers.

Would that make sense to you?

**Comment by benbeckingsale** [comment]
I don't think that it makes sense to classify a speaker as a `Voice` device because voice is not the primary mode of interaction.

Doing some research into Sonos useragents, I played some of my content using the Sonos app:

    Model: Play:1
    Version: 11.1 (build 56076060)
    Hardware Version: 1.20.1.6-2
    Series ID: A200

The resulting useragent was: `Linux UPnP/1.0 Sonos/56.0-76060 (ZPS12)`, very similar the useragents I brought up.

The speaker I used has no voice control, so to classify it as `DeviceClass: Voice` would not be right. In fact, it looks like just two Sonos speakers have voice support, whereas any of them could produce a similar useragent.

I hope to have some results for Sonos via Alexa in the next few days.
